Initial hwdec VAAPI support

lsiudut opened the pull request:
I really care about pushing it through, thus the PR even though I believe this
code needs few improvements.

When I was remote playing recently I noticed that every now and then the stream
shutters, especially during action / when there was many details on the screen.
Initially I thought this is stream bandwidth problem, but eventually I realized that
my ULP CPU got completely saturated. At first I wanted to hack on decoding threads
(it seems that there's only one created?) but using VAAPI seemed more elegant, so
here we are.

VAAPI is a interface used by Intel and AMD to accelerate video operations. Quite
widely supported by variety of apps, especially by ffmpeg/avconv which is used
in Chiaki. Enabling it's support was rather trivial and brings quite some improvements.

For needs of this change I had to modify VideoDecode construcor in order to toss
one more parameter - whether hw decode should be enabled or not. I don't feel like
passing just a boolean is very elegant, but definitely simpler than passing whole
settings object for example (especially that VideoDecode exists in context of
StreamSession where access to settings is aready limited). Also, I didn't want to
go as far as interfacing whole VideoDecode class...

Before we get to numbers there's one caveat - VAAPI produces results in NV12 color
space, so in order to display them in OpenGL widget used by Chiaki it must be converted
to YUV420P. Operation is not very fast, but from my quick benchmarks still faster than
software h264 decode (see the numbers below). As a next iteration I can try to interface
VideoDecode class and display widget and come up with VAAPI widget, which will use
VA surface to draw on (conversion can be skipped then, see `mpv` sources for example).

If it comes to numbers, here's `pidstat` session from PS4 main menu - I was simply
moving the cursor, enabling options to generate some movement:

First, no acceleration:
```
21:59:53      UID       PID    %usr %system  %guest   %wait    %CPU   CPU  Command
21:59:54     1000   2973781   20,00    6,00    0,00    0,00   26,00     6  chiaki
21:59:55     1000   2973781   36,00    6,00    0,00    0,00   42,00     5  chiaki
21:59:56     1000   2973781   35,00    8,00    0,00    0,00   43,00     0  chiaki
21:59:57     1000   2973781   35,00    5,00    0,00    0,00   40,00     4  chiaki
21:59:58     1000   2973781   30,00    5,00    0,00    1,00   35,00     4  chiaki
21:59:59     1000   2973781   36,00    6,00    0,00    0,00   42,00     0  chiaki
22:00:00     1000   2973781   38,00    4,00    0,00    0,00   42,00     4  chiaki
22:00:01     1000   2973781   31,00    5,00    0,00    0,00   36,00     3  chiaki
```

Then, with acceleration:
```
22:00:11     1000   2973781    9,00    6,00    0,00    0,00   15,00     1  chiaki
22:00:12     1000   2973781   17,00    7,00    0,00    1,00   24,00     3  chiaki
22:00:13     1000   2973781   19,00    6,00    0,00    0,00   25,00     2  chiaki
22:00:14     1000   2973781   17,00    7,00    0,00    0,00   24,00     2  chiaki
22:00:15     1000   2973781   18,00    7,00    0,00    0,00   25,00     1  chiaki
22:00:16     1000   2973781   18,00    6,00    0,00    0,00   24,00     7  chiaki
22:00:17     1000   2973781   17,00    8,00    0,00    0,00   25,00     0  chiaki
22:00:18     1000   2973781   18,00    6,00    0,00    0,00   24,00     5  chiaki
22:00:19     1000   2973781    2,00    1,00    0,00    0,00    3,00     3  chiaki
```

Generally CPU doesn't exceed 25% of single core when VAAPI is enabled, doesn't
matter on the scene dynamic.

lsiudut commented:
Need to add --enable-swscale to ffmpeg build, will do tomorrow.

dshine28D commented:
Boo mean peace of mind . Lean means running rich thats engine terms.
Trending is leaning in a hospital bed. Stream is the flow of a older
person's urinary flow. Fast means fast unless you ad as that's said yall
grew up with no slang or any sense of humor


On Sat, Apr 18, 2020, 5:34 PM Łukasz Siudut <notifications@github.com>
wrote:

> I really care about pushing it through, thus the PR even though I believe
> this
> code needs few improvements.
>
> When I was remote playing recently I noticed that every now and then the
> stream
> shutters, especially during action / when there was many details on the
> screen.
> Initially I thought this is stream bandwidth problem, but eventually I
> realized that
> my ULP CPU got completely saturated. At first I wanted to hack on decoding
> threads
> (it seems that there's only one created?) but using VAAPI seemed more
> elegant, so
> here we are.
>
> VAAPI is a interface used by Intel and AMD to accelerate video operations.
> Quite
> widely supported by variety of apps, especially by ffmpeg/avconv which is
> used
> in Chiaki. Enabling it's support was rather trivial and brings quite some
> improvements.
>
> For needs of this change I had to modify VideoDecode construcor in order
> to toss
> one more parameter - whether hw decode should be enabled or not. I don't
> feel like
> passing just a boolean is very elegant, but definitely simpler than
> passing whole
> settings object for example (especially that VideoDecode exists in context
> of
> StreamSession where access to settings is aready limited). Also, I didn't
> want to
> go as far as interfacing whole VideoDecode class...
>
> Before we get to numbers there's one caveat - VAAPI produces results in
> NV12 color
> space, so in order to display them in OpenGL widget used by Chiaki it must
> be converted
> to YUV420P. Operation is not very fast, but from my quick benchmarks still
> faster than
> software h264 decode (see the numbers below). As a next iteration I can
> try to interface
> VideoDecode class and display widget and come up with VAAPI widget, which
> will use
> VA surface to draw on (conversion can be skipped then, see mpv sources
> for example).
>
> If it comes to numbers, here's pidstat session from PS4 main menu - I was
> simply
> moving the cursor, enabling options to generate some movement:
>
> First, no acceleration:
>
> 21:59:53      UID       PID    %usr %system  %guest   %wait    %CPU   CPU  Command
> 21:59:54     1000   2973781   20,00    6,00    0,00    0,00   26,00     6  chiaki
> 21:59:55     1000   2973781   36,00    6,00    0,00    0,00   42,00     5  chiaki
> 21:59:56     1000   2973781   35,00    8,00    0,00    0,00   43,00     0  chiaki
> 21:59:57     1000   2973781   35,00    5,00    0,00    0,00   40,00     4  chiaki
> 21:59:58     1000   2973781   30,00    5,00    0,00    1,00   35,00     4  chiaki
> 21:59:59     1000   2973781   36,00    6,00    0,00    0,00   42,00     0  chiaki
> 22:00:00     1000   2973781   38,00    4,00    0,00    0,00   42,00     4  chiaki
> 22:00:01     1000   2973781   31,00    5,00    0,00    0,00   36,00     3  chiaki
>
> Then, with acceleration:
>
> 22:00:11     1000   2973781    9,00    6,00    0,00    0,00   15,00     1  chiaki
> 22:00:12     1000   2973781   17,00    7,00    0,00    1,00   24,00     3  chiaki
> 22:00:13     1000   2973781   19,00    6,00    0,00    0,00   25,00     2  chiaki
> 22:00:14     1000   2973781   17,00    7,00    0,00    0,00   24,00     2  chiaki
> 22:00:15     1000   2973781   18,00    7,00    0,00    0,00   25,00     1  chiaki
> 22:00:16     1000   2973781   18,00    6,00    0,00    0,00   24,00     7  chiaki
> 22:00:17     1000   2973781   17,00    8,00    0,00    0,00   25,00     0  chiaki
> 22:00:18     1000   2973781   18,00    6,00    0,00    0,00   24,00     5  chiaki
> 22:00:19     1000   2973781    2,00    1,00    0,00    0,00    3,00     3  chiaki
>
> Generally CPU doesn't exceed 25% of single core when VAAPI is enabled,
> doesn't
> matter on the scene dynamic.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/thestr4ng3r/chiaki/pull/205
> Commit Summary
>
>    - Initial hwdec VAAPI support
>
> File Changes
>
>    - *M* gui/CMakeLists.txt
>    <https://github.com/thestr4ng3r/chiaki/pull/205/files#diff-dbbdf13c20f044e206be8dfdaa6e8b36>
>    (4)
>    - *M* gui/include/settings.h
>    <https://github.com/thestr4ng3r/chiaki/pull/205/files#diff-2183027fc07f3cf0fc97e4aa5f0154ac>
>    (3)
>    - *M* gui/include/settingsdialog.h
>    <https://github.com/thestr4ng3r/chiaki/pull/205/files#diff-3ee3759f756225428d2e67d18e3304eb>
>    (2)
>    - *M* gui/include/streamsession.h
>    <https://github.com/thestr4ng3r/chiaki/pull/205/files#diff-33bc854fd71681239643ecfd362b8abb>
>    (1)
>    - *M* gui/include/videodecoder.h
>    <https://github.com/thestr4ng3r/chiaki/pull/205/files#diff-035ce8c8adeb8b49cc71ce76add6d38c>
>    (11)
>    - *M* gui/src/settingsdialog.cpp
>    <https://github.com/thestr4ng3r/chiaki/pull/205/files#diff-36f0de2434fea0344d5aeef6ad9dcccd>
>    (20)
>    - *M* gui/src/streamsession.cpp
>    <https://github.com/thestr4ng3r/chiaki/pull/205/files#diff-fa947f0c21803ee8562b6858c8974217>
>    (3)
>    - *M* gui/src/videodecoder.cpp
>    <https://github.com/thestr4ng3r/chiaki/pull/205/files#diff-f283cc02d743a876d1aed74e7daec068>
>    (91)
>
> Patch Links:
>
>    - https://github.com/thestr4ng3r/chiaki/pull/205.patch
>    - https://github.com/thestr4ng3r/chiaki/pull/205.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/thestr4ng3r/chiaki/pull/205>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOGTLNM5WH36CJYTABIDGVTRNIMHXANCNFSM4MLQO2VQ>
> .
>

thestr4ng3r commented:
Very nice!
Swscale is only needed for the nv12 -> yuv conversion, right? I think we should be able to do the conversion in the shader and pass the raw frame to it. The yuv isn't even the "native" format for displaying anyway, the shader has to convert it to rgb.

lsiudut commented:
Oh right, I see the GLSL code now! I'll play with it a bit (funny fact: it's in my todo list for at least 10 years).

Then I'll take a look on the failure in Android build, I don't really understand it.

Also, I just spent another 2h on gameplay and 15 min loadavg is 0.49, when previously it was going easily above 1. I also managed to eliminate all the wifi related lags. Flawless fun.

thestr4ng3r commented:
> Then I'll take a look on the failure in Android build, I don't really understand it.

Ignore the Android build, it's misconfiguration from my side. I was just too lazy until now to fix it for prs.

Cybereu commented:
Could compile and test on Ubuntu 19.10, and it's awesome !!! It`s clear how it become better especially over wifi.

lsiudut commented:
Glad to hear @Cybereu ! More improvements are coming, as per comment from @thestr4ng3r . I just had to educate myself, but now I think I know what to do, already started hacking on it.

In case I won't find enough time during the work week I'll land few more cleanups to this code and will suggest merging it as is, then prepare GLSL NV12 conversion as a separate PR.

lsiudut commented:
Got it. I do have working code already, just need to productize it. Doesn't seem to bring massive CPU improvements, but it's definitely more elegant.

Will update the PR tomorrow (UTC) time.

lsiudut commented:
Simple as that: https://gist.github.com/lsiudut/08bfbf542048abad4bc983b3ea983fa4

lsiudut commented:
Ok, just take a look on this abomination and tell me how to do it properly, so it doesn't haunt you at nights ;-).

cc @thestr4ng3r

lsiudut commented on gui/CMakeLists.txt:
```diff
@@ -20,7 +20,7 @@ if(WIN32)
 endif()
 
-find_package(FFMPEG REQUIRED COMPONENTS avcodec avutil)
+find_package(FFMPEG REQUIRED COMPONENTS avcodec avutil swscale)
```
Ah I can remove swscale already... Will do in the next iteration.

lsiudut commented on gui/include/avopenglwidget.h:
```diff
@@ -28,15 +28,36 @@ extern "C"
 
+struct PlaneConfig
+{
+	unsigned int width_divider;
+	unsigned int height_divider;
```
Silent assumption that most of formats will have additional panes with resolution equal or less than original frame.

lsiudut commented on gui/src/avopenglwidget.cpp:
```diff
@@ -93,6 +141,15 @@ AVOpenGLWidget::AVOpenGLWidget(VideoDecoder *decoder, QWidget *parent)
 	: QOpenGLWidget(parent),
 	decoder(decoder)
 {
+	for(auto &cc: conversion_configs)
+	{
+		if(decoder->PixelFormat() == cc.pixel_format)
+		{
+			conversion_config = &cc;
+			break;
+		}
+	}
+
```
I don't know how to return an error here if PixelFormat() is not found in available conversion configs. Just throw?

thestr4ng3r replied:
Yes, you can throw this: https://github.com/thestr4ng3r/chiaki/blob/master/gui/include/exception.h

thestr4ng3r commented on gui/src/avopenglwidget.cpp:
```diff
@@ -256,26 +309,31 @@ void AVOpenGLWidget::initializeGL()
+		frames[i].conversion_config = conversion_config;
+		f->glGenTextures(conversion_config->planes, frames[i].tex);
+		f->glGenBuffers(conversion_config->planes, frames[i].pbo);
+		uint32_t uv_default = 0x7f7f7f7f;
```
I don't think this is right for nv12, it shows up as dark green. Since it's more than one byte now, it should be better to do something like this to not run into endianness issues:
```c++
uint8_t uv_default[] = [0x7f, 0x7f];
```

lsiudut replied:
So funny fact - it worked before my rework. By then I actually assigned 127 << 8 | 127, what is effectively the same. I have no idea why it went dark-green.

Do you want me to iterate on it or shall we ignore it for now and just iterate?

thestr4ng3r commented on gui/src/videodecoder.cpp:
```diff
@@ -34,6 +39,35 @@ VideoDecoder::VideoDecoder(ChiakiLog *log) : log(log)
+		const char *hw_dec_eng = hardware_decode_engine_names[hw_decode_engine];
+		CHIAKI_LOGI(log, "Using hardware decode %s", hw_dec_eng);
+		type = av_hwdevice_find_type_by_name(hw_dec_eng);
+		if (type == AV_HWDEVICE_TYPE_NONE) {
```
remove the {} for single-line ifs

thestr4ng3r commented on gui/src/videodecoder.cpp:
```diff
@@ -34,6 +39,35 @@ VideoDecoder::VideoDecoder(ChiakiLog *log) : log(log)
+
+		for(int i = 0;; i++) {
+			const AVCodecHWConfig *config = avcodec_get_hw_config(codec, i);
+			if(!config) {
```
same

thestr4ng3r commented on gui/src/videodecoder.cpp:
```diff
@@ -34,6 +39,35 @@ VideoDecoder::VideoDecoder(ChiakiLog *log) : log(log)
+				throw VideoDecoderException("avcodec_get_hw_config failed");
+			}
+			if(config->methods & AV_CODEC_HW_CONFIG_METHOD_HW_DEVICE_CTX &&
+				config->device_type == type) {
```
Use Allman style
```suggestion
				config->device_type == type)
			{
```

lsiudut commented:
Out of curiosity - is there a reason why you allocate the texture buffer and memcpy into it instead of bind it directly to `frame->data[i]` in `glTexImage2D` call? Some data alignment issues? It seems to work for NV12 (but yeah, haven't tested YUV and now I'm too lazy to turn my PS4 on).

thestr4ng3r commented:
> Out of curiosity - is there a reason why you allocate the texture buffer and memcpy into it instead of bind it directly to `frame->data[i]` in `glTexImage2D` call? Some data alignment issues? It seems to work for NV12 (but yeah, haven't tested YUV and now I'm too lazy to turn my PS4 on).

Good question actually, it's been a while since I wrote that code but iirc it was primarily because of this else part:
```c++
if(frame->linesize[i] == width)
	memcpy(buf, frame->data[i], size);
else
{
	for(int l=0; l<height; l++)
		memcpy(buf + width * l * conversion_config->plane_configs[i].data_per_pixel, frame->data[i] + frame->linesize[i] * l, width * conversion_config->plane_configs[i].data_per_pixel);
}
```
Depending on the resolution (no hw acceleration + 540p), ffmpeg allocates a bit of padding after each horizontal line which `glTexImage2D()` couldn't handle. Without respecting it, you get something like this:
![Screenshot from 2020-04-22 09-57-32](https://user-images.githubusercontent.com/1460997/79955937-c3dd6180-847f-11ea-8b33-84bade51712e.png)

thestr4ng3r commented on gui/CMakeLists.txt:
```diff
@@ -79,7 +79,7 @@ if(CHIAKI_ENABLE_CLI)
 endif()
 
-target_link_libraries(chiaki FFMPEG::avcodec FFMPEG::avutil)
+target_link_libraries(chiaki FFMPEG::avcodec FFMPEG::avutil FFMPEG::swscale)
```
```suggestion
target_link_libraries(chiaki FFMPEG::avcodec FFMPEG::avutil)
```

thestr4ng3r commented on gui/include/videodecoder.h:
```diff
@@ -22,16 +22,32 @@
 extern "C"
 {
 #include <libavcodec/avcodec.h>
+#include <libswscale/swscale.h>
```
```suggestion
```

lsiudut replied:
Yea just noticed that, sorry. My local build of ffmpeg has it all so it was not failing. Fixing now.

lsiudut commented:
Ok, I just landed to minor things.

1) found the culprit why screen was dark green when `vaapi` is chosen - bad indexing of plane configs (`i` instead of `j`)
2) I changed conversion_config assignment to use designated initializers - it seems cleaner to me, but I have no idea if it fits the general spirit of the code, happy to roll that back (it's c++17 onwards AFAIK)

thestr4ng3r commented on gui/src/avopenglwidget.cpp:
```diff
@@ -256,26 +313,31 @@ void AVOpenGLWidget::initializeGL()
 			f->glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
 			f->glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
-			f->glTexImage2D(GL_TEXTURE_2D, 0, GL_R8, 1, 1, 0, GL_RED, GL_UNSIGNED_BYTE, j > 0 ? &uv_default : nullptr);
+			f->glTexImage2D(GL_TEXTURE_2D, 0, conversion_config->plane_configs[j].internal_format, 1, 1, 0, conversion_config->plane_configs[j].format, GL_UNSIGNED_BYTE, j > 0 ? &uv_default : nullptr);
```
Nice, you found the i/j issue yourself, I was just about to comment about it.
Just a small suggestion, better use the array directly here.
```suggestion
			f->glTexImage2D(GL_TEXTURE_2D, 0, conversion_config->plane_configs[j].internal_format, 1, 1, 0, conversion_config->plane_configs[j].format, GL_UNSIGNED_BYTE, j > 0 ? uv_default : nullptr);
```

thestr4ng3r commented:
> 2. I changed conversion_config assignment to use designated initializers - it seems cleaner to me, but I have no idea if it fits the general spirit of the code, happy to roll that back (it's c++17 onwards AFAIK)

Yeah please revert that, while it is indeed cleaner, I think the rest of the code should be compatible with c++11 (except some android-specific code).

lsiudut commented:
Ok, I think I addressed it all now. Btw I also removed vdpau option from settings dialog - I'll try to test it first and will put up another PR once done. May take a while though to find hardware that supports it.

Funnily enough there's fake vdpau engine that uses vaapi instead, but it's crashing my kernel. Another thing to waste my time on during The Great WFH.

Thanks for the reviews. It's been years since I was writing in C++ and I feel extremely rusted and insecure ;-).

thestr4ng3r commented:
Thank you very much!
My nvidia card is forwarded to a VM so I can't test vdpau myself currently either.
As you have already noted, loading the frames from the GPU to the CPU only to then push them into OpenGL may not be the ideal solution for presentation when using hw decoding, so there is potential to add more options later.

lsiudut commented:
Will do. BTW I also have a system with nvidia attached to VM, home made steambox. It runs windows currently though and underlying Linux is in headless mode.

Today after hours I'll setup another VM, this time Linux based, attach the card to it and will compile&test VDPAU there. So I'll be back :).

LiamDawe commented:
Would be great if this could expand to cover NVIDIA too, currently enabling the new option with an nvidia 1080 just reports this:

> Failed to initialize Stream Session: avcodec_get_hw_config failed

lsiudut commented:
Nvidia doesn't support vaapi per-se, but rather vdpau. Theoretically it's supported by ffmpeg, just need to run some tests to make sure that drawing won't get affected by not supported color space.

I'll definitely jump onto it in next days.

jeme81 commented:
Was there any progress on the Nvidia support? But for real even without the Nvidia support its so nice streaming PS4 to Linux. Thanks!
